### PR TITLE
Add triangular solve function for sparse CSR tensors in XPU

### LINKED ISSUE #3167: NotImplementedError: Could not run 'aten::triangular_solve.X' with arguments from the 'SparseCsrXPU' backend
Cases:
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int32_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_2_int64_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int32_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_False_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_block_triangular_solve_block_size_3_int64_noncontiguous_True_xpu_float64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_sparse_triangular_solve_xpu_complex128
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_sparse_triangular_solve_xpu_complex64
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_sparse_triangular_solve_xpu_float32
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_csr_xpu.TestSparseCSRXPU,test_sparse_triangular_solve_xpu_float64

```
NotImplementedError: Could not run 'aten::triangular_solve.X' with arguments from the 'SparseCsrXPU' backend. This could be because the operator doesn't exist for this backend, or was ...
```

### PR DESCRIPTION
**Pull request by tszulist-hbn**
Resolves: https://github.com/intel/torch-xpu-ops/issues/3167

This PR adds the SparseCsrXPU dispatch for triangular_solve by converting the sparse matrix to dense and delegating to the existing dense XPU triangular_solve implementation. This follows the same pattern used by other XPU sparse ops (addmm, baddbmm, etc.).

### Changes:
yaml/native/native_functions.yaml — Register SparseCsrXPU: triangular_solve_out_sparse_csr_xpu dispatch for both triangular_solve.X (structured out variant) and triangular_solve (structured delegate).
src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp — Implement triangular_solve_out_sparse_csr_xpu() which handles the zero-nnz edge case (fills X with NaN, matching CUDA behavior) and otherwise converts sparse A to dense before calling at::triangular_solve.

### Tests verified:
test_block_triangular_solve — 64/64 variants pass (block_size 2/3, int32/int64, contiguous/noncontiguous, float32/float64/complex64/complex128)
test_sparse_triangular_solve_xpu — 4/4 variants pass (float32/float64/complex64/complex128)

Created test_triangular_solve_sparse.py with test cases covering:

- **Upper/lower triangular** sparse CSR matrices
- **Multiple right-hand sides** (multi-column B)
- **Transpose** mode
- **Unit-triangular** mode
- **float32** precision (with relaxed tolerances)
- **Empty RHS** (exercises the `numel() == 0` early-exit path that fills with NaN)

Each test constructs a non-singular triangular matrix in CSR format, solves on XPU, and compares against the dense CPU result. The style follows the existing test_linalg.py pattern in the same directory.

### CODE REVIEW

**Review by guangyey — APPROVED** [review]
Overall LGTM. Could you please add a test case? Maybe `test/regressions` is a good place.
Defer to @CuiYifeng

File: src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp [hunk 1]
```diff
@@ -469,4 +470,23 @@ Tensor& addmv_out_sparse_compressed_xpu(
...
+    bool unitriangular,
+    Tensor& X,
+    Tensor& clone_A) {
+  if (B.numel() == 0 || X.numel() == 0 || A._nnz() == 0) {
+    X.fill_(NAN);
+    return std::tuple<Tensor&, Tensor&>(X, clone_A);
+  }
+  auto result =
+      at::triangular_solve(B, A.to_dense(), upper, transpose, unitriangular);
+  X.copy_(std::get<0>(result));
```
Comment (CuiYifeng) [inline src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp]: Please consider whether we can avoid this copy by using the out variant of `triangular_solve`. If so, we may need to add a `impl` function for `triangular_solve_out_sparse_csr_xpu`.

**Review by CuiYifeng — CHANGES_REQUESTED** [review]

File: src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp (hunk 1 above)
Comment (tszulist-hbn) [reply src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp]: @CuiYifeng, done.

**Review by CuiYifeng — COMMENTED** [review]
LGTM, but I noticed that `test_block_triangular_solve_block_size` like cases are still failed in CI https://github.com/intel/torch-xpu-ops/actions/runs/24331820702/job/71154120366?pr=3261. Please check, thanks.

File: src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp [hunk 2]
```diff
@@ -469,4 +470,25 @@ Tensor& addmv_out_sparse_compressed_xpu(
...
+    bool unitriangular,
+    Tensor& X,
+    Tensor& clone_A) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(A.is_sparse_csr());
+
+  if (B.numel() == 0 || X.numel() == 0 || A._nnz() == 0) {
+    X.fill_(NAN);
+    return std::tuple<Tensor&, Tensor&>(X, clone_A);
+  }
+  Tensor temp_clone_A = at::empty({0}, A.options().layout(at::kStrided));
```
Comment (CuiYifeng) [inline src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp:488]: Was the failure of `test_block_triangular_solve_block_size` like cases caused by the update of `clone_A`?

Comment (tszulist-hbn) [reply src/ATen/native/sparse/xpu/SparseCsrTensorMath.cpp:488]: It was. triangular_solve_out_sparse_csr_xpu was incorrectly passing clone_A to the dense triangular_solve_out function, causing clone_A to be filled when it should stay empty for sparse matrices. I updated it to use a temporary tensor, temp_clone_A, for the cloned coefficient.

**Review by CuiYifeng — APPROVED** [review]

**Comment by chuanqi129** [comment]
❌ Failed to merge PR: Repository rule violations found

A conversation must be resolved before this pull request can be merged.

**Comment by chuanqi129** [comment]
✅ PR has been successfully merged by @tszulist-hbn.
